client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx

```diff
@@ -373,16 +377,31 @@ export const PlanComparisonGrid: React.FC< PlanComparisonGridProps > = ( {
 	useEffect( () => {
 		let newVisiblePlans = displayedPlansProperties.map( ( { planName } ) => planName );
 
-		if ( isMediumBreakpoint ) {
-			newVisiblePlans = newVisiblePlans.slice( 0, 4 );
+		let visibleLength = newVisiblePlans.length;
+		if ( ! isInSignup ) {
+			visibleLength = isLargestBreakpoint ? 4 : visibleLength;
+			visibleLength = isLargeBreakpoint ? 3 : visibleLength;
 		}
 
-		if ( isSmallBreakpoint || ( isInSignup && isMediumBreakpoint ) ) {
-			newVisiblePlans = newVisiblePlans.slice( 0, 2 );
+		visibleLength = isMediumBreakpoint ? 2 : visibleLength;
+
+		if ( newVisiblePlans.length !== visibleLength ) {
+			const currentPlan = displayedPlansProperties.find( ( { current } ) => current );
+			if ( currentPlan ) {
+				newVisiblePlans.sort( ( a ) => ( a === currentPlan.planName ? -1 : 1 ) );
+			}
+			newVisiblePlans = newVisiblePlans.slice( 0, visibleLength );
 		}
 
 		setVisiblePlans( newVisiblePlans );
-	}, [ isMediumBreakpoint, intervalType ] );
+	}, [
+		isLargestBreakpoint,
+		isLargeBreakpoint,
+		isMediumBreakpoint,
+		intervalType,
+		planProperties,
```

Comment:
We need to add `planProperties` to ensure that switching between sites (Free vs Paid) updates the plan list.
Using `displayedPlansProperties`  seems to create an infinite loop.

Reply:
@chriskmnds is correct. This happens because of the code here.
https://github.com/Automattic/wp-calypso/blob/24a9d72f43ba6ba0a4fa32e756be040b3ddde31f/client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx#L363-L365

The filter continously creates a new array reference on each render, which triggers a call to this useEffect when displayedPlansProperties is in the deps array. The use effec also essentially set's a new array as state everytime it is triggered. This is what triggers the infinite loop.

Wrapping the [above line](https://github.com/Automattic/wp-calypso/blob/24a9d72f43ba6ba0a4fa32e756be040b3ddde31f/client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx#L363-L365) with [useMemo](https://beta.reactjs.org/reference/react/useMemo) should fix it.

Reply:
Good points and thanks for the suggestions!
useMemo seems to fix it.